parlai/nn/lr_scheduler.py

```diff
@@ -0,0 +1,520 @@
+
+        if opt.get('lr_scheduler') == 'none':
+            return None
+        elif decay == 1.0:
+            warn_once(
+                "Your LR decay is set to 1.0. Assuming you meant you wanted "
+                "to disable learning rate scheduling. Adjust --lr-scheduler-decay "
+                "if this is not correct."
+            )
+            return None
+        elif opt.get('lr_scheduler') == 'reduceonplateau':
+            scheduler = ReduceOnPlateauLRScheduler(
+                optimizer,
+                states,
+                hard_reset,
+                patience,
+                decay,
+                warmup_updates,
+                warmup_rate,
+            )
+        elif opt.get('lr_scheduler') == 'fixed':
+            scheduler = FixedLRScheduler(
+                optimizer,
+                states,
+                hard_reset,
+                patience,
+                decay,
+                warmup_updates,
+                warmup_rate,
+            )
+        elif opt.get('lr_scheduler') == 'invsqrt':
+            scheduler = InvSqrtLRScheduler(
+                optimizer,
+                states,
+                hard_reset,
+                patience,
+                decay,
+                warmup_updates,
+                warmup_rate,
+                invsqrt_lr_decay_gamma,
+            )
+        elif opt.get('lr_scheduler') == 'cosine':
+            scheduler = CosineLRScheduler(
+                optimizer,
+                states,
+                hard_reset,
+                patience,
+                decay,
+                warmup_updates,
+                warmup_rate,
+                max_lr_steps,
+            )
+        elif opt.get('lr_scheduler') == 'linear':
+            scheduler = LinearLRScheduler(
+                optimizer,
+                states,
+                hard_reset,
+                patience,
+                decay,
+                warmup_updates,
+                warmup_rate,
+                max_lr_steps,
+            )
+        else:
+            raise ValueError(
+                "Don't know what to do with --lr-scheduler '{}'".format(
+                    opt.get('lr_scheduler')
+                )
+            )
+
+        # time to load LR state from the checkpoint, if possible.
+        if (
+            # there is already an old LR scheduler saved on disk
+            states
+            # and the old LR scheduler is different
+            and states.get('lr_scheduler_type') != opt['lr_scheduler']
+            # and we're not already using a fresh scheduler
+            and not hard_reset
+        ):
+            # the LR scheduler changed, start things fresh
+            warn_once("LR scheduler is different from saved. Starting fresh!")
+            hard_reset = True
+
+        if not hard_reset:
+            scheduler.load_state(states)
+            # do the actual loading (if possible)
+
+        scheduler.init_warmup_scheduler(
```

Comment:
i'm wondering if this shouldn't be private; it's not something you'd want to expose to TorchAgent

Reply:
(also maybe you can get rid of passing the unnecessary args here? i.e. all of these you pass to the scheduler in the constructor?)